### PR TITLE
Dev 1.1.0

### PR DESCRIPTION
**Pull request by bclements**
# Comprehensive Metrics Agent Refactor and Enhancement

## Summary
This PR represents a major overhaul of the Vega Metrics Agent, implementing direct metric collection from Kubernetes APIs and Kubelet, enhancing collection capabilities, and improving overall performance and reliability.

### Architecture Changes
- Removed dependency on metrics-server
- Implemented direct metric collection from:
  - Kubernetes APIs
  - Kubelet API endpoints (/metrics, /stats, /metrics/cadvisor)
  - Node stats endpoints
  - Container runtime metrics
- Added support for collecting additional metrics:
  - Detailed node resource metrics
  - Pod resource utilization
  - Container statistics
  - Kubelet health metrics
  - Custom and external metrics
  - Hardware topology information
  - Process and runtime metrics
  - Volume and filesystem metrics

### Performance and Concurrency
- Implemented concurrent metric collection with goroutines
- Added mutex protection for thread-safe metric updates
- Optimized collection intervals
- Improved error handling with detailed error reporting
- Added retry logic for failed metric collections
- Enhanced memory management for large clusters
- Implemented efficient metric batching

### Security Enhancements
- Enhanced service account token handling
- Added verification of collector client identity
- Improved TLS configuration for kubelet API communication
- Added secure transport configuration options
- Enhanced RBAC permissions with principle of least privilege
- Added support for secure and insecure kubelet endpoints
- Improved secret handling and environment variable security

### API and Permission Updates
- Enhanced ClusterRole with comprehensive permissions for:
  - Core API resources
  - Workload resources (batch and apps)
  - Networking resources
  - Metrics APIs (metrics.k8s.io, custom.metrics.k8s.io, external.metrics.k8s.io)
  - Storage and CSI resources
  - Security resources
  - Node resources and topology
  - GPU metrics (nvidia.com)
- Added access to non-resource URLs for Kubelet metrics
- Added permissions for self-subject reviews

### Metric Collection Improvements
- Parallel collection of:
  - CPU metrics (usage, throttling, scheduling)
  - Memory metrics (usage, cgroup stats)
  - Network metrics (bandwidth, packets, errors)
  - Disk I/O metrics
  - System metrics (load, uptime)
  - Volume metrics (usage, capacity)
  - Container stats (resource usage, state)
  - Hardware topology (CPU, memory layout)
  - Process metrics (counts, states)
  - Runtime metrics (container runtime stats)
  - Filesystem metrics (usage, inodes)

### Development and Build Improvements
- Updated Go version from 1.22 to 1.23
- Version bump from 1.0.0 to 1.1.0-rc1
- Added new Makefile targets:
  - `alldev`: Development build target with security checks
  - `alldevnosec`: Development build target without security checks
  - `docker-build-dev` for local development
- Updated dependencies to latest versions
- Enhanced debug logging capabilities
- Improved error handling and retry logic

### Helm Chart Enhancements
- Added initialization container to wait for permissions
- Implemented Helm hook weights for proper deployment ordering
- Added development-specific gitignore for values-dev.yaml
- Enhanced service account configuration
- Added annotations for deployment dependencies
- Improved resource requests and limits
- Added support for custom environment variables

### Configuration and Environment
- Added KUBERNETES_CLIENT_LOG_LEVEL environment variable
- Added default VEGA_INSECURE setting for development environments
- Enhanced configuration validation
- Added support for:
  - Custom collection intervals
  - Metric filtering
  - Debug logging levels
  - TLS configuration
  - Proxy settings
  - Resource limits

## Testing Notes
- Verify metric collection works without metrics-server
- Test direct Kubelet API communication
- Validate all new metric types are being collected
- Ensure proper RBAC permissions in restricted environments
- Test in both development and production configurations
- Verify concurrent collection performance
- Test error handling and retry logic
- Validate memory usage in large clusters

### Breaking Changes
- Metrics-server is no longer required
- New RBAC permissions must be reviewed and approved
- Direct Kubelet access needs to be enabled
- Update Go toolchain to 1.23 for development

### Configuration Updates
- Review and update any custom metric collection settings
- Verify Kubelet API access in your environment
- Check security implications of expanded RBAC roles
- Update monitoring dashboards for new metrics
- Review resource limits and requests

## Performance Impact
- Improved metric collection efficiency through direct API access
- Reduced dependency chain and potential points of failure
- More granular metric collection capabilities
- Optimized memory usage through efficient data structures
- Enhanced concurrent collection reduces overall collection time

## Documentation Updates
- Updated README with new configuration options
- Added debugging guide
- Updated deployment instructions
- Added security considerations
- Updated API documentation

## Future Considerations
- Support for additional metric types
- Enhanced filtering capabilities
- Custom metric aggregation
- Advanced scheduling options
- Additional cloud provider integrations

### CODE REVIEW

File: bin/vega-metrics-agent
Comment (Jameswood-vega) [inline bin/vega-metrics-agent:1]: did you mean to add the bin?

File: charts/vega-metrics-agent/Chart.yaml [hunk 1]
```diff
@@ -10,12 +10,6 @@
 apiVersion: v2
 name: vega-metrics-agent
 description: A Helm chart for deploying Vega Metrics Agent
-version: 1.0.0
-appVersion: "1.0.0"
+version: 1.1.0-dev
```
Comment (Jameswood-vega) [inline charts/vega-metrics-agent/Chart.yaml]: should this be -dev?

File: charts/vega-metrics-agent/charts/metrics-server-3.12.1.tgz
Comment (Jameswood-vega) [inline charts/vega-metrics-agent/charts/metrics-server-3.12.1.tgz:1]: did you mean to add binary?

File: charts/vega-metrics-agent/values.yaml [hunk 2]
```diff
@@ -13,7 +13,7 @@ serviceAccount:
   name: vega-metrics-sa
 image:
   repository: public.ecr.aws/c0f8b9o4/vegacloud/vega-metrics-agent
-  tag: 1.0.0
+  tag: 1.1.0-rc
```
Comment (Jameswood-vega) [inline charts/vega-metrics-agent/values.yaml]: should this be 1.1.0-rc1

Comment (Jameswood-vega) [reply charts/vega-metrics-agent/values.yaml]: i had to add the 1 to make it work.

**Review by Jameswood-vega — APPROVED** [review]

File: bin/vega-metrics-agent
Comment (bclements) [reply bin/vega-metrics-agent:1]: I've added a .gitignore now and fix.

**Pull request merged**